onnxruntime/core/mlas/lib/power/qgemm_kernel_power10.cpp

```diff
@@ -0,0 +1,1199 @@
+        Mval = 4;
+    }
+    while(CountN > 0) {
+        const int8_t *a = A;
+        typedef __vector unsigned char vec_t;
+        const uint8_t *b = B;
+        int32_t *C1;
+        __vector_quad acc0, acc1, acc2, acc3, acc4, acc5, acc6, acc7;
+        //
+        // Initialize the accumulators with zero.
+        //
+        __builtin_mma_xxsetaccz(&acc0);
+        __builtin_mma_xxsetaccz(&acc1);
+        __builtin_mma_xxsetaccz(&acc2);
+        __builtin_mma_xxsetaccz(&acc3);
+        __builtin_mma_xxsetaccz(&acc4);
+        __builtin_mma_xxsetaccz(&acc5);
+        __builtin_mma_xxsetaccz(&acc6);
+        __builtin_mma_xxsetaccz(&acc7);
+        MLAS_INT32X4 result[4] = {0};
+        MLAS_INT32X4 result1[4] = {0};
+        size_t k = PackedCountK*4;
+        size_t k1 = PackedCountK;
+        //
+        // Compute the output block using POWER10 MMA builtins.
+        //
+        while (k >= 16) {
+            vec_t *va = const_cast<vec_t *>(reinterpret_cast<const vec_t *>(&a[0]));
+            vec_t *vb = const_cast<vec_t *>(reinterpret_cast<const vec_t *>(&b[0]));
+            __builtin_mma_xvi8ger4pp (&acc0, va[0], vb[0]);
+            __builtin_mma_xvi8ger4pp (&acc0, va[1], vb[1]);
+            __builtin_mma_xvi8ger4pp (&acc0, va[2], vb[2]);
+            __builtin_mma_xvi8ger4pp (&acc0, va[3], vb[3]);
+            if (CountM >= 8) {
+                __builtin_mma_xvi8ger4pp (&acc4, va[4], vb[0]);
+                __builtin_mma_xvi8ger4pp (&acc4, va[5], vb[1]);
+                __builtin_mma_xvi8ger4pp (&acc4, va[6], vb[2]);
+                __builtin_mma_xvi8ger4pp (&acc4, va[7], vb[3]);
+            }
+            vb = const_cast<vec_t *>(reinterpret_cast<const vec_t *>(&b[k1*16]));
+            __builtin_mma_xvi8ger4pp (&acc1, va[0], vb[0]);
+            __builtin_mma_xvi8ger4pp (&acc1, va[1], vb[1]);
+            __builtin_mma_xvi8ger4pp (&acc1, va[2], vb[2]);
+            __builtin_mma_xvi8ger4pp (&acc1, va[3], vb[3]);
+            if (CountM >= 8) {
+                __builtin_mma_xvi8ger4pp (&acc5, va[4], vb[0]);
+                __builtin_mma_xvi8ger4pp (&acc5, va[5], vb[1]);
+                __builtin_mma_xvi8ger4pp (&acc5, va[6], vb[2]);
+                __builtin_mma_xvi8ger4pp (&acc5, va[7], vb[3]);
+            }
+            vb = const_cast<vec_t *>(reinterpret_cast<const vec_t *>(&b[k1*32]));
+            __builtin_mma_xvi8ger4pp (&acc2, va[0], vb[0]);
+            __builtin_mma_xvi8ger4pp (&acc2, va[1], vb[1]);
+            __builtin_mma_xvi8ger4pp (&acc2, va[2], vb[2]);
+            __builtin_mma_xvi8ger4pp (&acc2, va[3], vb[3]);
+            if (CountM >= 8) {
+                __builtin_mma_xvi8ger4pp (&acc6, va[4], vb[0]);
+                __builtin_mma_xvi8ger4pp (&acc6, va[5], vb[1]);
+                __builtin_mma_xvi8ger4pp (&acc6, va[6], vb[2]);
+                __builtin_mma_xvi8ger4pp (&acc6, va[7], vb[3]);
+            }
+            vb = const_cast<vec_t *>(reinterpret_cast<const vec_t *>(&b[k1*48]));
```

Comment:
>vec_t *>(&b[k1*48])); [](http://example.com/codeflow?start=60&length=21)

how much perf gain do you see comparing with simply processing CountN by 4?

Reply:
I have not separately evaluated performance differences with CountN by 4. Overall compared to the generic case, we see up to 20X improvements on POWER10.